chore(rfc): APM stats support in Datadog source (#9900)
vectordotdev/vector: rfcs/2021-11-03-9862-ingest-apm-stats-along-traces-in-dd-agent-source.md

> # RFC  - 2021-11-XX - Ingest APM stats in `datadog-agent` source with consistent user experience

Reviewer: Can we update the title to follow the pattern? (issue ID and date)

> Datadog traces support in `datadog-agent` source was initially documented in this [RFC][trace-support-pr]. However the
> Datadog `trace-agent` submit more data that juste traces. It also send statistics ("APM Stats") about running time of
> each instrumented resource (i.e. a given piece of code) that are aggregated over time (based on 100% of the traces
> received). Those APM stats are very important as they can highlight code hot spots and ease aggregation. Also as an
> integral part of the Datadog APM product, APM stats should be handled by Vector.

Reviewer: ```suggestion
Datadog traces support in the `datadog_agent` source was initially documented in this [RFC][trace-support-pr]. However, the
Datadog `trace-agent` submits more data than just traces. It also sends statistics ("APM Stats") about the running time of
each instrumented resource (i.e. a given piece of code) that are aggregated over time (based on 100% of the traces
received). Those APM stats are very important as they highlight code hot spots and ease aggregation. And, as a result, APM stats should be handled by Vector.
```

[trace-support-pr]: https://github.com/vectordotdev/vector/pull/9634

## Context

> Traces are new in Vector, initial support is focusing on traces coming from the Datadog Agent, the
> [RFC][trace-support-pr] discussing about traces describes how those traces will be handled in Vector and highlighted the
> need to support APM stats.

Reviewer: ```suggestion
Traces are new in Vector, initial support focuses on traces coming from the Datadog Agent, the
[RFC][trace-support-pr] discussing traces describes how those traces will be handled in Vector and highlights the
need to support APM stats.
```

> APM stats are encoded using protobuf, the current schema is available in the [datadog-agent][apm-stats-proto]
> repository. APM stats may be computed by the tracing lib in some cases or most of the time by the trace agent, but once
> they are emitted by the `trace-agent` there is no difference except a [boolean value][client-side-marker] that indicates
> where it was computed. They are sent by the `trace-agent` to the same endpoint as trace payloads, only the path differ,
> allowing easy discrimination between trace & APM stats payloads.

Reviewer: ```suggestion
APM stats are encoded using Protobuf, the current schema is available in the [datadog-agent][apm-stats-proto]
repository. APM stats may be computed by the tracing lib in some cases, or most of the time by the `trace-agent`, but once
they are emitted by the `trace-agent` there is no difference except a [boolean value][client-side-marker] that indicates
where it was computed. They are sent by the `trace-agent` to the same endpoint as trace payloads, only the paths differ,
allowing easy discrimination between trace & APM stats payloads.
```

> Those stats are computed by a component named [concentrator] in the trace-agent. There is a [dedicated
> path][client-stats-aggregator] for APM stats that comes directly from tracing libraries. But ultimately they flow
> through datadog with the same [sending code][stats-writer].

Reviewer: ```suggestion
Those stats are computed by a component named [concentrator] in the `trace-agent`. There is a [dedicated
path][client-stats-aggregator] for APM stats that comes directly from tracing libraries. But ultimately they flow
through Datadog with the same [sending code][stats-writer].
```

> If we details a bit the actual content of a stats payload, it is an aggregate of
> [`ClientGroupedStats`][client-grouped-stats-proto]:

Reviewer: ```suggestion
Below is an example of a stats payload, it is an aggregate of [`ClientGroupedStats`][client-grouped-stats-proto]:
```

```protobuf
string service = 1;
string name = 2;
string resource = 3;
uint32 HTTP_status_code = 4;
string type = 5;
string DB_type = 6; // db_type might be used in the future to help in the obfuscation step
uint64 hits = 7; // count of all spans aggregated in the groupedstats
uint64 errors = 8; // count of error spans aggregated in the groupedstats
uint64 duration = 9; // total duration in nanoseconds of spans aggregated in the bucket
bytes okSummary = 10; // ddsketch summary of ok spans latencies encoded in protobuf
bytes errorSummary = 11; // ddsketch summary of error spans latencies encoded in protobuf
bool synthetics = 12; // set to true on spans generated by synthetics traffic
uint64 topLevelHits = 13; // count of top level spans aggregated in the groupedstats
```

> So it is basically a group of various metrics that can be represented in Vector (Sketches are support since this
> [PR][sketch-pr]). In the proto definition sketches are stored as unstructured bytes slices, but those
> [fields][call-to-toproto] are filled with a [protobuf encoded ddsketch][ddsktech-proto]. Given that sketches in Vector
> are also [heavily based on ddsketch][vector-sketch], APM stats sketches can be converted to/from the Vector internal
> representation without incuring too much accuracy losses, but this would require significant work to implement those
> conversion.

Reviewer: ```suggestion
As you can see, it is a group of various metrics that can be represented in Vector as such (Sketches are supported since this
[PR][sketch-pr]). In the proto definition sketches are stored as unstructured bytes slices, but those
[fields][call-to-toproto] are filled with a [protobuf encoded ddsketch][ddsktech-proto]. Given that sketches in Vector
are also [heavily based on ddsketch][vector-sketch], APM stats sketches can be converted to/from the Vector internal
representation without incurring too much accuracy loss, but this would require significant work to implement that
conversion.
```

> conversion.

Reviewer 2: Do you know if the accuracy loss is bounded and to what, if so?

This opens two major very different paths for APM stats in Vector:

> * Either the APM stats are emitted as a log, each [`ClientGroupedStats`][client-grouped-stats-proto] would be mapped to
>   one log event.
> * Or they are emitted as metrics each value from every [`ClientGroupedStats`][client-grouped-stats-proto] would be
>   emitted as a metric with all upper level information store as tags, but as each metric event flows independently from
>   other ones this would imply adding a significant re-aggregation logic in the traces sink.

Reviewer: ```suggestion
1. Either the APM stats are emitted as a log, each [`ClientGroupedStats`][client-grouped-stats-proto] would be mapped to
  one log event.
2. Or they are emitted as metrics. Each value from every [`ClientGroupedStats`][client-grouped-stats-proto] would be
  emitted as a metric with all upper level information stored as tags. Each metric would flow independently from
  others and this would be require significant re-aggregation logic in the `datadog_traces` sink.
```

>   other ones this would imply adding a significant re-aggregation logic in the traces sink.

Reviewer 2: Can you clarify a bit about what re-aggregation would need to happen?

Author: APM stats are a group of metrics as shown below

```
string service = 1;
string name = 2;
string resource = 3;
uint32 HTTP_status_code = 4;
string type = 5;
string DB_type = 6; // db_type might be used in the future to help in the obfuscation step
uint64 hits = 7; // count of all spans aggregated in the groupedstats
uint64 errors = 8; // count of error spans aggregated in the groupedstats
uint64 duration = 9; // total duration in nanoseconds of spans aggregated in the bucket
bytes okSummary = 10; // ddsketch summary of ok spans latencies encoded in protobuf
bytes errorSummary = 11; // ddsketch summary of error spans latencies encoded in protobuf
bool synthetics = 12; // set to true on spans generated by synthetics traffic
uint64 topLevelHits = 13; // count of top level spans aggregated in the groupedstats
```

So to match the Vector metric datamodel we would have to generate several metrics from the above protobuf payloads: several counters (hits/errors/topLevelHits), a gauge (duration) and two sketches (okSummary & errorSummary). All those metrics will bear all other information ecnoded as metrics tags and once they reach the datadog_traces sink they will need to be re-aggregated to recreate a working APM stats payload. Aggregation should account for all dimension of a single [original payload](https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/pb/stats.proto#L10-L16).

>   other ones this would imply adding a significant re-aggregation logic in the traces sink.

Reviewer 2: They seem to be submitted independently. From the above:

>  They are sent by the `trace-agent` to the same endpoint as trace payloads, only the path differ,
allowing easy discrimination between trace & APM stats payloads.

So they are separate requests. I am curious if the delay between submission matters though.

>   other ones this would imply adding a significant re-aggregation logic in the traces sink.

Author: For sure there can be a small delay as the `trace-agent` itself delays stats by 10sec to account for possibly late traces (as per https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/stats/concentrator.go#L174-L176). This more a tradeoff between buffering long enough not to miss anything and the delay for the data to be available in the APM app.

> Those to approach can be mixed up together if we introduce one of the following ability to bundle multiple metrics into
> a single event:

Reviewer: ```suggestion
Those two approaches can be mixed together if we introduce one of the following abilities to bundle multiple metrics into
a single event:
```

> * Allow multiple metric samples into a single metric event

Author: This would make a lot of sense, If we can get all metrics from each `ClientGroupedStats` into a single batch for each of those `ClientGroupedStats`, this would greatly simplify the `datadog_traces` sink implementation.

> * Allow log events to embed metrics, it would mean adding a `Metric` type to the `Value` enum

Author: I agree this one is probably a bad idea. I think we can dismissed it.

> This also raises the question of having a second (assuming that the `datadog-agent` sources accepts Datadog Agent
> metrics - [RFC][dd-agent-metric-rfc]), unrelated, metric stream coming out of the `datadog-agent` source. Obviously
> event ingested representing APM stats will have to be routed along with traces, and most often they will follow a
> different path that other plain metrics/logs received from a core agent. Thus it is suggested to re-arrange the
> `datadog-agent` source, many option are available (Additional details on what exactly are "Datadog Agents" can be
> found in the trace support [RFC][trace-support-pr] and may provide relevant context for undermentionned points):

Reviewer: ```suggestion
This also raises the question of having a second (assuming that the `datadog_agent` sources accepts Datadog Agent
metrics - [RFC][dd-agent-metric-rfc]), unrelated, metric stream coming out of the `datadog_agent` source. The
Vector event ingested representing APM stats will have to be routed along with traces, and most often they will follow a
different path that other plain metrics/logs received from a `core-agent`. Thus it is suggested to re-arrange the
`datadog_agent` source, many option are available (Additional details on what exactly are "Datadog Agents" can be
found in the trace support [RFC][trace-support-pr] and may provide relevant context for undermentionned points):
```

> * Keeping a single `datadog-agent` source:

Author: I agree, that why I put other solution in the alternative section. I'll come up with a more detailed implementation proposal for those "named" outputs a source (it could apply to transform too) ASAP.

  * and add a settings to switch between agent kind: `agent: <TYPE>` where `<TYPE>`
  could be  `core` (would support metrics & logs - we could optionnally add `logs` & `metrics` to only allow logs or
  metrics, along with `core` that would allow both), `trace` and could be extended to support `process`, `security` and
  so on

>   * or introduce the ability for a source to have multiple output for the `datadog-agent` source, e.g. <SRC_ID>.metrics, <SRC_ID>.logs, <SRC_ID>.traces, <SRC_ID>.apm_stats, etc.

Reviewer: This is my preference. When a user defines a `datadog_agent` source it accepts _all_ data and outputs over the various channels. I like the `apm_stats` channel.

Reviewer 2: Likewise. I prefer the multiple outputs.

> * Keeping a single `datadog-agent` source:
>   * and add a settings to switch between agent kind: `agent: <TYPE>` where `<TYPE>`
>   could be  `core` (would support metrics & logs - we could optionnally add `logs` & `metrics` to only allow logs or
>   metrics, along with `core` that would allow both), `trace` and could be extended to support `process`, `security` and
>   so on
>   * or introduce the ability for a source to have multiple output for the `datadog-agent` source, e.g. <SRC_ID>.metrics, <SRC_ID>.logs, <SRC_ID>.traces, <SRC_ID>.apm_stats, etc.
> * Another solution would be to keep one Vector source per kind of Datadog Agent an then we would have the following
>   Vector sources:
>   * `datadog-core` or the current `datadog-agent` sources that would receive the data sent by the Datadog "core" Agent
>     (the agent that collects logs & metrics).
>   * `datadog-trace` that would support all data sent by the `trace-agent`
>   * And so on as the support list grows:
>     * `datadog-process` source for the Datadog `process-agent`
>     * `datadog-security` for the Datadog security Agent
>     * Etc.

Reviewer: ```suggestion
* Keeping a single `datadog_agent` source:
  * and add a settings to switch between agent kind: `agent: <TYPE>` where `<TYPE>`
  could be  `core` (would support metrics & logs - we could optionally add `logs` & `metrics` to only allow logs or
  metrics, along with `core` that would allow both), `trace` and could be extended to support `process`, `security` and
  so on
  * or introduce the ability for a source to have multiple output for the `datadog-agent` source, e.g. <SRC_ID>.metrics, <SRC_ID>.logs, <SRC_ID>.traces, <SRC_ID>.apm_stats, etc.
* Another solution would be to keep one Vector source per kind of Datadog Agent an then we would have the following
  Vector sources:
  * `datadog_core` or the current `datadog_agent` sources that would receive the data sent by the Datadog "core" Agent
    (the agent that collects logs & metrics).
  * `datadog_trace` that would support all data sent by the `trace-agent`
  * And so on as the support list grows:
    * `datadog_process` source for the Datadog `process-agent`
    * `datadog_security` for the Datadog security Agent
    * Etc.
```

    * Etc.

[apm-stats-proto]: https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/pb/stats.proto
[client-side-marker]: https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/pb/stats.proto#L15
[concentrator]: https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/stats/concentrator.go
[client-stats-aggregator]: https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/stats/client_stats_aggregator.go#L25-L45
[stats-writer]: https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/writer/stats.go
[client-grouped-stats-proto]: https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/pb/stats.proto#L55-L70
[sketch-pr]: https://github.com/vectordotdev/vector/pull/9178
[call-to-toproto]: https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/stats/statsraw.go#L53-L62
[ddsktech-proto]: https://github.com/DataDog/sketches-go/blob/0829b5a6a3c9627d1faee6c690611067ee110e38/ddsketch/pb/ddsketch.proto
[vector-sketch]: https://github.com/vectordotdev/vector/blob/4db23b3/lib/vector-core/src/metrics/ddsketch.rs#L202-L213
[dd-agent-metric-rfc]: https://github.com/vectordotdev/vector/pull/9057

> ## Cross cutting concerns

Reviewer: This is another place I would mention the upcoming schemas support for interoperability.

Author: Will do.

# RFC  - 2021-11-03 - Ingest APM stats in `datadog-agent` source with consistent user experience

Datadog traces support in the `datadog_agent` source was initially documented in this [RFC][trace-support-pr]. However, the

> Datadog `traceagent` submits more data than just traces. It also sends statistics ("APM Stats") about the running time of

Reviewer 3: ```suggestion
Datadog `trace-agent` submits more data than just traces. It also sends statistics ("APM Stats") about the running time of
```
?

    * And every 10 seconds (this is the sending interval of the trace-agent) serializing and flushing those to Datadog.
      To account for late metrics the sink would have to keep 2 or 3 buckets in the past and delay flushing accordingly.
      This would rely on the [bucket timestamp][btime] kept by the trace agent and [stored in APM stats
      payload][csb-start].

At a certain point in time, when [sketches-rs][sketches-rs] will be production ready:

* Swich Vector sketches to use the crate [sketches-rs][sketches-rs] instead of the Agent based implementation
* Then relocate conversion logic: sketch conversion will then be effectively useless in traces handling, but the
  `datadog_agent` source and the `datadog_metrics` sink will then have to handle conversion between vector sketches
  (plain ddsketches) and the agent variation (note: it's on the crate roadmap to offer that kind of conversion)

[btime]: https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/stats/concentrator.go#L148-L159 [csb-start]:
https://github.com/DataDog/datadog-agent/blob/dc2f202/pkg/trace/pb/stats.proto#L47

## Rationale

* We should Keep valuable metrics relevant to end-user

> * Dropping APM stats would cause current user to loose some insight on execution time.

Reviewer 3: ```suggestion
* Dropping APM stats would cause current user to lose some insight on execution time.
```
